src/powershell/Program.cs

```diff
@@ -19,7 +20,232 @@ public sealed class ManagedPSEntry
         /// </param>
         public static int Main(string[] args)
         {
+#if UNIX
+            if (HasLoginSpecified(args, out int loginIndex))
+            {
+                return ExecPwshLogin(args, loginIndex);
+            }
+#endif
             return UnmanagedPSEntry.Start(string.Empty, args, args.Length);
         }
+
+#if UNIX
+        /// <summary>
+        /// Checks args to see if -Login has been specified.
+        /// </summary>
+        /// <param name="args">Arguments passed to the program.</param>
+        /// <param name="loginIndex">The arg index (in argv) where -Login was found.</param>
+        /// <returns></returns>
+        private static bool HasLoginSpecified(string[] args, out int loginIndex)
+        {
+            loginIndex = -1;
+
+            // Parameter comparison strings, stackalloc'd for performance
+            for (int i = 0; i < args.Length; i++)
+            {
+                string arg = args[i];
+
+                // Must look like '-<name>'
+                if (arg == null || arg.Length < 2 || arg[0] != '-')
+                { 
+                    continue;
+                }
+
+                // Check for "-Login" or some prefix thereof
+                if (IsParam(arg, "login", "LOGIN"))
+                {
+                    loginIndex = i;
+                    return true;
+                }
+
+                // After -File and -Command, all parameters are passed
+                // to the invoked file or command, so we can stop looking.
+                if (IsParam(arg, "file", "FILE")
+                    || IsParam(arg, "command", "COMMAND"))
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
```

Comment:
Should we cache `arg.length`?

Reply:
It's just a field on the string/array, and one I imagine the CLR is well optimised to retrieve and recognise calls to

Reply:
I would not hope for it - with TC many optimizations are skipped.

Reply:
https://github.com/dotnet/coreclr/blob/48ff0937552e540f21835391b693daf47ffabece/src/System.Private.CoreLib/src/System/String.CoreCLR.cs#L46-L57